nilearn/tests/test_surface.py

```diff
@@ -0,0 +1,393 @@
+                  mesh, np.eye(4), img.shape, mask=np.ones((3, 3, 2)))
+
+
+def test_sampling_affine():
+    # check sampled (projected) values on a toy image
+    img = np.ones((4, 4, 4))
+    img[1, :, :] = 2
+    nodes = [[1, 1, 2], [10, 10, 20], [30, 30, 30]]
+    mesh = [np.asarray(nodes), None]
+    affine = 10 * np.eye(4)
+    affine[-1, -1] = 1
+    texture = surface._nearest_voxel_sampling(
+        [img], mesh, affine=affine, radius=1, kind='ball')
+    assert_array_equal(texture[0], [1., 2., 1.])
+    if LooseVersion(scipy.__version__) < LooseVersion('0.14'):
+        return
+    texture = surface._interpolation_sampling(
+        [img], mesh, affine=affine, radius=0, kind='ball')
+    assert_array_almost_equal(texture[0], [1.1, 2., 1.])
+
+
+def test_sampling():
+    mesh = _flat_mesh(5, 7, 4)
+    img = _z_const_img(5, 7, 13)
+    mask = np.ones(img.shape, dtype=int)
+    mask[0] = 0
+    projectors = [surface._nearest_voxel_sampling]
+    if LooseVersion(scipy.__version__) >= LooseVersion('0.14'):
+        projectors.append(surface._interpolation_sampling)
+    for kind in ('line', 'ball'):
+        for projector in projectors:
+            projection = projector([img], mesh, np.eye(4),
+                                   kind=kind, radius=0.)
+            assert_array_almost_equal(projection.ravel(), img[:, :, 0].ravel())
+            projection = projector([img], mesh, np.eye(4),
+                                   kind=kind, radius=0., mask=mask)
+            assert_array_almost_equal(projection.ravel()[7:],
+                                      img[1:, :, 0].ravel())
+            assert_true(np.isnan(projection.ravel()[:7]).all())
+
+
+def test_vol_to_surf():
+    # test 3d niimg to cortical surface projection and invariance to a change
+    # of affine
+    mni = datasets.load_mni152_template()
+    mesh = _generate_surf()
+    _check_vol_to_surf_results(mni, mesh)
+    if LooseVersion(nb.__version__) > LooseVersion('1.2.0'):
+        fsaverage = datasets.fetch_surf_fsaverage5().pial_left
+        _check_vol_to_surf_results(mni, fsaverage)
+
+
+def _check_vol_to_surf_results(img, mesh):
+    for kind in ['ball', 'line']:
+        for interpolation in ['linear', 'nearest']:
+            proj_1 = surface.vol_to_surf(
+                img, mesh, kind=kind, interpolation=interpolation)
+            assert_true(proj_1.ndim == 1)
+            img_rot = image.resample_img(
+                img, target_affine=rotation(np.pi / 3., np.pi / 4.))
+            proj_2 = surface.vol_to_surf(
+                img_rot, mesh, kind=kind, interpolation=interpolation)
+            # The projection values for the rotated image should be very close
+            # to the projection for the original image
+            diff = np.abs(proj_1 - proj_2) / np.abs(proj_1)
+            assert_true(np.mean(diff[diff < np.inf]) < .03)
+            img_4d = image.concat_imgs([img, img])
+            proj_4d = surface.vol_to_surf(
```

Comment:
same here, can you please test with mask ?

Reply:
I understand that you have tested masked_indices but I would like to test with vol_to_surf ?